internal/handlers/common/killcursors.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
+	"github.com/FerretDB/FerretDB/internal/clientconn/cursor"
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/handlers/commonparams"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
+// KillCursors is a part of common implementation of the killCursors command.
+func KillCursors(ctx context.Context, msg *wire.OpMsg, registry *cursor.Registry) (*wire.OpMsg, error) {
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	command := document.Command()
+
+	db, err := GetRequiredParam[string](document, "$db")
+	if err != nil {
+		return nil, err
+	}
+
+	collection, err := GetRequiredParam[string](document, command)
+	if err != nil {
+		return nil, err
+	}
+
+	username, _ := conninfo.Get(ctx).Auth()
+
+	cursors, err := GetRequiredParam[*types.Array](document, "cursors")
+	if err != nil {
+		return nil, err
+	}
+
+	iter := cursors.Iterator()
+	defer iter.Close()
+
+	cursorsKilled := types.MakeArray(0)
+	cursorsNotFound := types.MakeArray(0)
+	cursorsAlive := types.MakeArray(0)
+	cursorsUnknown := types.MakeArray(0)
+
+	for {
+		i, v, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				break
+			}
+
+			return nil, lazyerrors.Error(err)
+		}
+
+		id, ok := v.(int64)
+		if !ok {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
```

Comment:
If cursors array contains a valid cursor and an invalid cursor like `bson.A{Long(1234),"foo"}`, I *feel* the command shouldn't kill the first cursor. :thinking:

Reply:
Added a test for that also demonstrates the usage of XFail

Reply:
Great, very fun of this helper function!